File: binarytree/codes/traversal.cpp
#include <iostream>
#include <vector>
#include <string>
#include <queue>

using namespace std;

struct Node
{
    int data;
    Node* left = nullptr;
    Node* right = nullptr;
    Node(int data){
       this->data = data;
    }
};


Node* createTree()
{
    Node* root = new Node(1);
    root->left = new Node(2);
    root->left->left = new Node(4);
    root->left->right = new Node(5);
    root->right = new Node(3);
    
    return root;
}


void preOrder(Node* root)
{
    cout << root->data;
    if (root->left != NULL){
        preOrder(root->left);
    }
    if (root->right != NULL){
        preOrder(root->right);
    }
}

void postOrder(Node* root)
{
    if (root->left != NULL){
        postOrder(root->left);
    }
    if (root->right != NULL){
        postOrder(root->right);
    }
    cout << root->data;
}

void inOrder(Node* root)
{
    if (root->left != NULL){
        inOrder(root->left);
    }
    cout << root->data;
    if (root->right != NULL){
        inOrder(root->right);
    }
}

void levelOrder(Node* root)
{
    queue< Node* > q;
    if( root!=NULL )
    {
        q.push(root);
    }
    while(q.size() != 0)
    {
        cout << q.front()->data;
        if (q.front()->left != NULL)
        {
            q.push(q.front()->left);
        }
        if (q.front()->right != NULL)
        {
            q.push(q.front()->right);
        }
        q.pop();
    }
}

int main()
{
   //-> is a shorthand for (*x).field, where x is a pointer to a variable of type struct account
   Node* root = createTree();

   preOrder(root);
   cout << "******************" << endl;
   postOrder(root);
   cout << "******************" << endl;
   inOrder(root);
   cout << "******************" << endl;
   levelOrder(root);
   cout << "******************" << endl;
}

File: binarytree/codes/traversal.py
import queue

class TreeNode:
  def __init__(self, data, left=None, right=None):
    self.data = data
    self.left = left
    self.right = right

class Traversal:
  def preOrder(self, root):
    print(root.data)
    if root.left is not None:
      self.preOrder(root.left)
    if root.right is not None:
      self.preOrder(root.right)

  def inOrder(self, root):
    if root.left is not None:
      self.inOrder(root.left)
    print(root.data)
    if root.right is not None:
      self.inOrder(root.right)

  def postOrder(self, root):
    if root.left is not None:
      self.postOrder(root.left)
    if root.right is not None:
      self.postOrder(root.right)
    print(root.data)

  def levelOrder(self, root):
    q = queue.Queue()
    if root is not None:
      q.put(root)
    else:
      return
    while not q.empty():
      tempt = q.get()
      print(tempt.data)
      if tempt.left is not None:
        q.put(tempt.left)
      if tempt.right is not None:
        q.put(tempt.right)


def main():
  root = TreeNode('f')
  nodeb = TreeNode('b')
  nodeg = TreeNode('g')
  nodea = TreeNode('a')
  noded = TreeNode('d')
  nodei = TreeNode('i')
  nodec = TreeNode('c')
  nodee = TreeNode('e')
  nodeh = TreeNode('h')

  root.left = nodeb
  root.right = nodeg
  nodeb.left = nodea
  nodeb.right = noded
  noded.left = nodec
  noded.right = nodee
  nodeg.right = nodei
  nodei.left = nodeh

  tv = Traversal()
  # tv.preOrder(root)
  # tv.inOrder(root)
  # tv.postOrder(root)
  tv.levelOrder(root)


if __name__ == '__main__':
    main()
